src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java

```diff
@@ -0,0 +1,90 @@
+import org.seqdoop.hadoop_bam.SAMRecordWritable;
+import scala.Tuple2;
+
+import java.io.IOException;
+
+/**
+ * ReadsSparkSink writes GATKReads to a file. This code lifts from the HadoopGenomics/Hadoop-BAM
+ * read writing code as well as from bigdatagenomics/adam.
+ */
+public class ReadsSparkSink {
+
+    // We need an output format for saveAsNewAPIHadoopFile.
+    public static class MyOutputFormat extends KeyIgnoringBAMOutputFormat<NullWritable> {
+        public static SAMFileHeader bamHeader = null;
+
+        public static void setHeader(final SAMFileHeader header) {
+            bamHeader = header;
+        }
+
+        @Override
+        public RecordWriter<NullWritable, SAMRecordWritable> getRecordWriter(TaskAttemptContext ctx) throws IOException {
+            setSAMHeader(bamHeader);
+            return super.getRecordWriter(ctx);
+        }
+    }
+
+    /**
+     * writeParallelReads writes rddReads to outputFile with header as the file header.
+     * @param ctx the JavaSparkContext to write.
+     * @param outputFile path to the output bam.
+     * @param rddReads reads to write.
+     * @param header the header to put at the top of the file.s
+     */
+    public static void writeParallelReads(
+            final JavaSparkContext ctx, final String outputFile, final JavaRDD<GATKRead> rddReads,
+            final SAMFileHeader header) throws IOException {
+        // Set the header on the main thread.
+        MyOutputFormat.setHeader(header);
+
+        // The expected format for writing is JavaPairRDD where the key is ignored and the value is a
+        // SAMRecordWritable. We use GATKRead as the key, so we can sort the reads before writing them to a file.
+        JavaPairRDD<GATKRead, SAMRecordWritable> rddSamRecordWriteable = rddReads.mapToPair(gatkRead -> {
+            SAMRecord samRecord = gatkRead.convertToSAMRecord(header);
+            SAMRecordWritable samRecordWritable = new SAMRecordWritable();
+            samRecordWritable.set(samRecord);
+            return new Tuple2<>(gatkRead, samRecordWritable);
+        });
+
+        final JavaPairRDD<GATKRead, SAMRecordWritable> out =
+                rddSamRecordWriteable.sortByKey(new ReadCoordinateComparator(header)).coalesce(1); // needed?
+
+        // MyOutputFormat is a static class, so we need to copy the header to each worker then call
+        // MyOutputFormat.setHeader.
+        final Broadcast<SAMFileHeader> broadcast = ctx.broadcast(header);
+
+        final JavaPairRDD<GATKRead, SAMRecordWritable> finalOut = out.mapPartitions(tuple2Iterator -> {
+            MyOutputFormat.setHeader(broadcast.getValue());
+            return new IteratorIterable<>(tuple2Iterator);
+        }).mapToPair(t -> t);
+
+        finalOut.saveAsNewAPIHadoopFile(outputFile, GATKRead.class, SAMRecordWritable.class, MyOutputFormat.class);
+        // At this point, the file is in outputFile/part-r-0000. We move it out of that dir and into it's own well-named
```

Comment:
The file could be in 'part-m-0000' if written from a job that doesn't have a reduce stage. This won't generalize beyond one file (i.e. for the sharded case). You should be able to avoid all this file moving by overriding `getDefaultWorkFile` in your output format, but it's a bit tricky to get right. You could try something like the following to get a feel for it:

```
  @Override
  public Path getDefaultWorkFile(TaskAttemptContext context,
      String extension) throws IOException {
    FileOutputCommitter committer =
        (FileOutputCommitter) format.getOutputCommitter(context);
    return new Path(committer.getWorkPath(), "foo.bam"); // in reality get filename from a property you set on the configuration
  }
```